desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,882 @@
+    weight : float, ndarray, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        len(weight) must be equal to Objective.dim_f
+    normalize : bool
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate error at.
+    field_grid : Grid, optional
+        Grid used to discretize ext_field.
+    name : str
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _linear = False
+    _print_value_fmt = "Boundary Error: {:10.3e} "
+    _units = "(T)"
+    _coordinates = "rtz"
+
+    def __init__(
+        self,
+        eq,
+        ext_field,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        grid=None,
+        field_grid=None,
+        name="Vacuum boundary error",
+    ):
+        if target is None and bounds is None:
+            target = 0
+        self._grid = grid
+        self._ext_field = ext_field
+        self._field_grid = field_grid
+        things = [eq]
+
+        super().__init__(
+            things=things,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        if self._grid is None:
+            grid = LinearGrid(
+                rho=np.array([1.0]),
+                M=eq.M_grid,
+                N=eq.N_grid,
+                NFP=int(eq.NFP),
+                sym=False,
+            )
+        else:
+            grid = self._grid
+
+        pres = np.max(np.abs(eq.compute("p")["p"]))
+        curr = np.max(np.abs(eq.compute("current")["current"]))
+        errorif(
+            not np.all(grid.nodes[:, 0] == 1.0),
+            ValueError,
+            "grid contains nodes not on rho=1",
+        )
+        warnif(
+            pres > 1e-8,
+            UserWarning,
+            f"Pressure is non-zero (max {pres} Pa), "
+            + "VacuumBoundaryError will be incorrect.",
+        )
+        warnif(
+            curr > 1e-8,
+            UserWarning,
+            f"Current is non-zero (max {curr} A), "
+            + "VacuumBoundaryError will be incorrect.",
+        )
+
+        self._eq_data_keys = [
+            "B",
+            "R",
+            "zeta",
```

Comment:
should this be `phi`, not `zeta`? For when general toroidal angle is implemented `zeta` won't be the same as `phi` but we need this to be the toroidal cylindrical angle `phi` for `compute_magnetic_field`

Reply:
The singular integral stuff current assumes zeta==phi. It could be generalized but I'd have to think about it more so I just kept everything as zeta for now.